src/errorHandler/errorHandler.ts

```diff
@@ -1,68 +1,143 @@
-import { GraphQLError } from 'graphql';
-import { ApolloError } from 'apollo-server-errors';
+import {
+  GraphQLError,
+  GraphQLFormattedError,
+  GraphQLErrorOptions,
+} from 'graphql';
+import { unwrapResolverError } from '@apollo/server/errors';
+
+/**
+ * Internally managed error codes.  If a new error is added here,
+ * it should be added to `NO_REPORT_ERRORS` in `sentryPlugin` if
+ * we do not want to report it to sentry.
+ */
+export enum InternalErrorCode {
+  BAD_USER_INPUT = 'BAD_USER_INPUT',
+  FORBIDDEN = 'FORBIDDEN',
+  INTERNAL_SERVER_ERROR = 'INTERNAL_SERVER_ERROR',
+  NOT_FOUND = 'NOT_FOUND',
+  UNAUTHENTICATED = 'UNAUTHENTICATED',
+}
 
 /**
  * Used for formatting errors returned to the client. Hide any
  * errors that might reveal server details. Handle special cases
  * that we want to use to provide more information to the client
  * (e.g. NotFoundError).
  */
-export function errorHandler(error: GraphQLError): GraphQLError {
-  if (error.originalError instanceof NotFoundError) {
-    return new GraphQLNotFoundError(error);
-  } else if (
-    error instanceof ApolloError ||
-    error.originalError instanceof ApolloError
-  ) {
+export function errorHandler(
+  formattedError: GraphQLFormattedError,
+  error: unknown
+): GraphQLFormattedError {
+  if (unwrapResolverError(error) instanceof GraphQLError) {
     // Keep GraphQL errors intact
     // e.g. failed parsing, bad input
-    return error;
+    return formattedError;
   } else {
     // Mask other kinds of errors
-    return new InternalServerError(error);
+    return new InternalServerError('Internal server error', {
+      originalError: error as Error,
+      // preserve path and locations
```

Comment:
Do we want to keep these? Preserved here for test compatibility, but could be revealing to somebody trying to intentionally cause errors.

Reply:
yeah can you clarify what the risk is? I think the utility of it is to know which part of the query is associated with the error

Reply:
we can prototype with one microservie e.g annotations-api and see how we want to organize this

Reply:
I don't know if it is a risk that is on our threat model, or that we want to consider as a part of our threat model. That is why I am asking. It seems like a detail you might want to obfuscate, similar to turning off introspection. Errors are almost always more expensive to handle than happy path code flows. In this context, we don't really know what this error is, how bad the outcome is, or how expensive it is to handle it. If somebody is fuzzing our service to try to find an expensive query or mutation, this is useful information to them. It might be that we accept that, and it might also be that now is not the right time to redact this.

If we do want it redacted eventually, and need an alternate solution for debugging, it might be good to implement an end-to-end tracking identifier, so that clients / we can get an exact stack trace for obfuscated errors from sentry more easily.